### ISSUE TITLE
Document how to use tokio-uds with tokio-proto

### ISSUE THREAD

**Issue by nicholasbishop** [tokio-rs/tokio-uds#5]
The tokio documentation has a simple example of setting up a TCP service with tokio-proto: https://tokio.rs/docs/getting-started/simple-server/#configure-and-run

I'm trying to understand how to do something similar with unix sockets, but it's not clear to me how to connect my implementation of a `Service` with a `UnixStream`.

**Comment by alexcrichton** [comment]
Sounds reasonable to me! The [`TcpServer`](https://github.com/tokio-rs/tokio-proto/blob/master/src/tcp_server.rs) code is actually pretty simple and basically just boils down to [this function](https://github.com/tokio-rs/tokio-proto/blob/d53f6c77b493398466852e240ae03d5dbe5684c2/src/tcp_server.rs#L127-L150). The trick is the `BindServer` trait which typically takes any I/O object, including a `UnixStream`.

**Comment by nicholasbishop** [comment]
The documentation for [BindServer](https://docs.rs/tokio-proto/0.1.0/tokio_proto/trait.BindServer.html) says "This trait is not intended to be implemented directly". Is this an exception to that rule, or should one of the other traits suggested in that documentation be used?

**Comment by alexcrichton** [comment]
@nicholasbishop oh yeah it's not implemented directly, but it can be called with a UnixStream. Most implementations of it (transitively through the impl you see) are generic over `T: Io`

**Comment by nicholasbishop** [comment]
I think I am still lost :) I was able to make a modified copy of `tcp_server.rs` that serves over a unix socket, but I don't feel any closer to understanding _why_ `tcp_server.rs` is implemented the way it is.

When you say "it's not implemented directly, but it can be called with a UnixStream", does "it" mean the `BindServer` trait? If so, what does it mean to call the trait with a `UnixStream`?

At a higher level, what is the point of the `BindServer` trait, as opposed to a more direct `Server` trait?

Actually, where are the other four traits mentioned in `BindServer`'s documentation? It refers to `pipeline::Server` and three others, but the `pipeline` module doesn't document a trait of that name.

**Comment by alexcrichton** [comment]
There are a number of main "proto" traits

* [`pipeline::ServerProto`](https://docs.rs/tokio-proto/0.1.0/tokio_proto/pipeline/trait.ServerProto.html)
* [`multiplex::ServerProto`](https://docs.rs/tokio-proto/0.1.0/tokio_proto/multiplex/trait.ServerProto.html)
* [`streaming::pipeline::ServerProto`](https://docs.rs/tokio-proto/0.1.0/tokio_proto/streaming/pipeline/trait.ServerProto.html)
* [`streaming::multiplex::ServerProto`](https://docs.rs/tokio-proto/0.1.0/tokio_proto/streaming/multiplex/trait.ServerProto.html)

These are intended to be implemented by protocol authors, and work in the various styles as described by the modules.

The [`BindServer`](https://docs.rs/tokio-proto/0.1.0/tokio_proto/trait.BindServer.html) trait then has a number of blanket impls to implement itself for free on your protocol once you implement one or more of the above Proto traits. The purpose of `BindServer` is to instantiate the protocol internals for you, e.g. the pipeline/multiplex dispatch.

The [`bind_server`](https://docs.rs/tokio-proto/0.1.0/tokio_proto/trait.BindServer.html#tymethod.bind_server) method takes any I/O object  and spawns a future that handles the connection. This I/O object can be a unix stream.

The tl;dr; of `TcpServer` is:

* Create a `Core`
* Create a `TcpListener`
* Crate a loop over each accepted connection
* For each accepted connection, call `bind_server`

Does that help clear things up?

**Comment by dgriffen** [comment]
This is a bit of a late response, but I just published a [crate](https://crates.io/crates/tokio-uds-proto) that makes a `UnixServer` and `UnixClient` with a similar API to `tokio-proto`

**Comment by alexcrichton** [comment]
Thanks @dgriffen!

**Comment by nicholasbishop** [comment]
Nice! I have my tokio experimentation on the back burner but I'll definitely give it a try when I get back to it. Thanks for putting that crate up.

**Comment by Dash83** [comment]
Has there been any progress on this issue? I currently have a project that uses a nanomsg as the networking library since I need to switch between TCP and UDS connections for different scenarios. 

However, I'm looking to move away from that library for several reasons and Tokio in general looks fantastic for my purposes, I just wanted to make sure I can create UDS connections without problems first, and I have yet to find an example of how this works.

Any documentation you could point me to?
